scrapy/pipelines/media.py

```diff
@@ -141,9 +140,11 @@ def _compatible(self, func):
         def wrapper(*args, **kwargs):
             if self._expects_item[func.__name__]:
                 return func(*args, **kwargs)
-            else:
+            elif 'item' in kwargs:
                 kwargs.pop('item', None)
-                return func(*args, **kwargs)
+            elif is_item(args[-1]):
+                args = args[:-1]
```

Comment:
I’m not sure about these 2 lines. `is_item` does not return `True` when the input _is_ an item, only when it _may be_ one.

What if we add `item` as a [keyword-only parameter](https://python-3-for-scientists.readthedocs.io/en/latest/python3_advanced.html#keyword-only-arguments) to those methods? Then we only need to worry about `kwargs` again.

Reply:
Yep we could do that, it would remove the uncertainty around `item` being passed as a positional/keyworded argument and make handling much easier! Thanks for the link. 
Will update the docs and warning message as well!